tests/codegen/vec_pop_push_noop.rs

```diff
@@ -0,0 +1,25 @@
+// compile-flags: -O
+
+#![crate_type = "lib"]
+
+#[no_mangle]
+pub fn noop(v: &mut Vec<u8>) {
+    // CHECK: start:
+    // CHECK-NEXT: getelementptr
+    // CHECK-NEXT: load
+    // CHECK-NEXT: icmp eq
+    // CHECK-NEXT: br
+
+    // CHECK: add
+    // CHECK-NEXT: getelementptr
+    // CHECK-NEXT: load
+    // CHECK-NEXT: icmp
+    // CHECK-NEXT: tail call
```

Comment:
Hmm, the way this is written it feels a bit too specific, in that lots of little changes could make it fail (like `bb3` getting a different name), but also that this `tail call` could be to something that reallocates and it'd pass.

Can you maybe focus it in a bit on what the critical part is?

Perhaps have another function in here showing that `push` normally can call `__rust_realloc`, but that this one doesn't?

A possible starting point:
```rust
#[no_mangle]
// CHECK-LABEL: @noop(
pub fn noop(v: &mut Vec<u8>) {
    // CHECK-NOT: __rust_alloc
    // CHECK-NOT: __rust_realloc
    // CHECK-NOT: call
    // CHECK: tail call void @llvm.assume
    // CHECK-NOT: __rust_alloc
    // CHECK-NOT: __rust_realloc
    // CHECK-NOT: call
    if let Some(x) = v.pop() {
        v.push(x)
    }
}

#[no_mangle]
// CHECK-LABEL: @push_byte(
pub fn push_byte(v: &mut Vec<u8>) {
    // CHECK: call {{.+}} @__rust_alloc(
    // CHECK: call {{.+}} @__rust_realloc(
    v.push(3);
}
```

Reply:
Hmm, maybe it'll have to look at `reserve_for_push` instead of the allocation functions directly...

Reply:
Hi,
I see your point but wouldn't it then be better to go for a codegen test on the assembly itself which is empty?

Reply:
No, because everything should be codegen tests where possible -- random permutations of temporary assembly registers and differing calling conventions make assembly tests something that's a last resort.